janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/IndexQueryBuilder.java

```diff
@@ -152,6 +184,73 @@ public String getUnknownKeyName() {
     // Builder Methods
     //################################################
 
+    /**
+     * Given constraints and result type, construct a direct index query
+     * If no direct index query can be built, return NULL
+     *
+     * @param optProfiler
+     * @param constraints
+     * @param resultType
+     * @return
+     */
+    public IndexQueryBuilder constructIndex(QueryProfiler optProfiler,
+                                            List<PredicateCondition<String, JanusGraphElement>> constraints,
+                                            ElementCategory resultType) {
+        optProfiler.startTimer();
+        // Prepare constraints
+        final MultiCondition<JanusGraphElement> conditions = QueryUtil.constraints2QNF(tx, constraints);
+
+        final Set<Condition> coveredClauses = new HashSet<>();
+        final IndexSelectionStrategy.SelectedIndexQuery selectedIndex = indexSelector.selectIndices(
+            resultType, conditions, coveredClauses, OrderList.NO_ORDER, serializer);
+        if (selectedIndex.isEmpty()) {
+            return null;
+        }
+        JointIndexQuery query = selectedIndex.getQuery();
+
+        assert query.size() == 1;
+        assert query.getQuery(0).getIndex().isMixedIndex() == true;
+
+        setIndex(query.getQuery(0).getIndex().getName());
+        setResultType(resultType);
+
+        StringBuilder queryBuilder = new StringBuilder();
+        constructQuery(queryBuilder, constraints.get(0));
+        for (int i = 1; i < constraints.size(); i++) {
+            queryBuilder.append(" AND ");
+            constructQuery(queryBuilder, constraints.get(i));
+        }
+
+        setQuery(queryBuilder.toString());
+        optProfiler.stopTimer();
+        return this;
+    }
+
+    private void constructQuery(final StringBuilder queryBuilder,
+                                final PredicateCondition<String, JanusGraphElement> constraint) {
+        if (ElementCategory.VERTEX.equals(resultType)) {
+            setPrefixInternal(VERTEX_PREFIX);
+            queryBuilder.append('v');
+        } else {
+            assert ElementCategory.EDGE == resultType;
+            setPrefixInternal(EDGE_PREFIX);
+            queryBuilder.append('e');
+        }
+        queryBuilder.append(".\"" + constraint.getKey() + "\":");
+        if (Contain.IN.equals(constraint.getPredicate())) {
+            queryBuilder.append("(");
+            for (Object value : (List) constraint.getValue()) {
+                queryBuilder.append("\"" + value + "\"");
+            }
+            queryBuilder.append(")");
+        } else if (Text.CONTAINS.equals(constraint.getPredicate()) || Cmp.EQUAL.equals(constraint.getPredicate())) {
+            queryBuilder.append("\"" + constraint.getValue() + "\"");
+        } else {
+            // TODO: we should support all types as long as a single mixed query can be constructed
```

Comment:
Now I think constructing a direct index query might be the wrong direction. Using existing `Stream<String> query(IndexQuery query, KeyInformation.IndexRetriever information, BaseTransaction tx)` in IndexProvider and appending `count` seems better.